### PR TITLE
Bugfix for algorithm in get_distance()

### PR DESCRIPTION
**Pull request by djunkins**
Updated algorithm use spherical law of cosines to calculate spherical
distance in get_distance() to address issue found in pull request #2068 

Reference to the algorithm at:

	https://www.movable-type.co.uk/scripts/latlong.html

Signed-off-by: Doug Junkins <junkins@foghead.com>

### Describe the pull request:

- [x] Bug fix
- [ ] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Related issues:
#2068

### CODE REVIEW

**Review by dirkhh — APPROVED** [review]
Thanks - I'll wait for Travis out of principle, but this appears to fix the bug I reported elsewhere.

**Comment by bstoeger** [comment]
> https://www.movable-type.co.uk/scripts/latlong.html

According to this page, the old algorithm was completely wrong! Here's the corrected version according to this page:

```
diff --git a/core/divesite.c b/core/divesite.c
index b5ce99a8d..956defaa1 100644
--- a/core/divesite.c
+++ b/core/divesite.c
@@ -77,12 +77,13 @@ struct dive_site *get_dive_site_by_gps_and_name(char *name, const location_t *lo
 // Calculate the distance in meters between two coordinates.
 unsigned int get_distance(const location_t *loc1, const location_t *loc2)
 {
+	double lat1_r = udeg_to_radians(loc1->lat.udeg);
 	double lat2_r = udeg_to_radians(loc2->lat.udeg);
 	double lat_d_r = udeg_to_radians(loc2->lat.udeg - loc1->lat.udeg);
 	double lon_d_r = udeg_to_radians(loc2->lon.udeg - loc1->lon.udeg);
 
 	double a = sin(lat_d_r/2) * sin(lat_d_r/2) +
-		cos(lat2_r) * cos(lat2_r) * sin(lon_d_r/2) * sin(lon_d_r/2);
+		cos(lat1_r) * cos(lat2_r) * sin(lon_d_r/2) * sin(lon_d_r/2);
 	double c = 2 * atan2(sqrt(a), sqrt(1.0 - a));
 
 	// Earth radious in metres
```
Of course, this still has problems with floating point imprecision for antipodes. So one would still have to clamp `a` into the 0..1.0 range.

I think this deserves a CHANGELOG entry, as it fixes a quite severe bug...?

**Comment by bstoeger** [comment]
I benchmarked both versions with the following two files:

test2.cpp:
```
#include <cmath>

static inline double udeg_to_radians(int udeg)
{
	return (udeg * M_PI) / (1000000.0 * 180.0);
}

int do_haver(int lat1, int lon1, int lat2, int lon2)
{
	double lat1_r = udeg_to_radians(lat1);
	double lat2_r = udeg_to_radians(lat2);
	double lat_d_r = udeg_to_radians(lat2 - lat1);
	double lon_d_r = udeg_to_radians(lon2 - lon1);

	double a = sin(lat_d_r/2) * sin(lat_d_r/2) +
		cos(lat1_r) * cos(lat2_r) * sin(lon_d_r/2) * sin(lon_d_r/2);
	if (a > 1.0) a = 1.0;
	if (a < 0.0) a = 0.0;

	double c = 2 * atan2(sqrt(a), sqrt(1.0 - a));

	return lrint(6371000 * c);
}

int do_cos(int lat1, int lon1, int lat2, int lon2)
{
	double lat1_r = udeg_to_radians(lat1);
	double lat2_r = udeg_to_radians(lat2);
	double lon1_r = udeg_to_radians(lon1);
	double lon2_r = udeg_to_radians(lon2);
	double c = acos((sin(lat1_r) * sin(lat2_r)) +
		(cos(lat1_r) * cos(lat2_r) * cos(lon1_r - lon2_r)));

	return lrint(6371000 * c);
}
```

test.cpp:
```
#include <boost/timer/timer.hpp>
#include <cstdint>
#include <cstdlib>
#include <ctime>
#include <iostream>

int do_haver(int lat1, int lon1, int lat2, int lon2);
int do_cos(int lat1, int lon1, int lat2, int lon2);

static uint64_t bench_haver(int lat1, int lon1, int lat2, int lon2)
{
	boost::timer::auto_cpu_timer t;

	uint64_t res = 0;
	for (int i = 0; i < 10000000; ++i)
		res += do_haver(lat1, lon1, lat2, lon2);
	return res;
}

static uint64_t bench_cos(int lat1, int lon1, int lat2, int lon2)
{
	boost::timer::auto_cpu_timer t;

	uint64_t res = 0;
	for (int i = 0; i < 10000000; ++i)
		res += do_cos(lat1, lon1, lat2, lon2);
	return res;
}

static int rand_udeg()
{
	return rand() % 1000000; // Yes, that's not a good way to generate random numbers!
}

int main()
{
	srand(time(0));
	int lat1 = rand_udeg();
	int lon1 = rand_udeg();
	int lat2 = rand_udeg();
	int lon2 = rand_udeg();
	uint64_t res_haver = bench_haver(lat1, lon1, lat2, lon2);
	uint64_t res_cos = bench_cos(lat1, lon1, lat2, lon2);
	std::cout << res_haver << ' ' << res_cos << '\n';
}
```
Compiled with `g++ -Wall -O3 test.cpp test2.cpp -lboost_timer`.
Turns out with a few exceptions the haversine version is significantly faster. Typical run:
```
$ ./a.out 
 0.684676s wall, 0.680000s user + 0.000000s system = 0.680000s CPU (99.3%)
 0.787008s wall, 0.790000s user + 0.000000s system = 0.790000s CPU (100.4%)
407810000000 407810000000
```

But sometimes the situation is reversed...?

File: core/divesite.c [hunk 1]
```diff
@@ -77,13 +77,13 @@ struct dive_site *get_dive_site_by_gps_and_name(char *name, const location_t *lo
...
+	double lon1_r = udeg_to_radians(loc1->lon.udeg);
+	double lon2_r = udeg_to_radians(loc2->lon.udeg);
 
-	double a = sin(lat_d_r/2) * sin(lat_d_r/2) +
-		cos(lat2_r) * cos(lat2_r) * sin(lon_d_r/2) * sin(lon_d_r/2);
-	double c = 2 * atan2(sqrt(a), sqrt(1.0 - a));
+	double c = acos((sin(lat1_r) * sin(lat2_r)) +
+		(cos(lat1_r) * cos(lat2_r) * cos(lon1_r - lon2_r)));
 
 	// Earth radious in metres
```
Comment (bstoeger) [inline core/divesite.c]: While touching this file, perhaps fix the typo in "radious"?

**Comment by djunkins** [comment]
I must be more jet lagged than I thought, because I looked at that algorithm several times and missed that one of the terms was wrong. Based on your benchmarks above, I will correct the Haversine function and add the bounding tests, add a note to CHANGELOG.md and resubmit (oh, and fix the spelling of radius.)

**Comment by Subsurface-divelog** [comment]
This pull request **introduces 2 alerts** when merging 23e660c5e6ffaf18c1ddbb5f7387aca455ee7ad0 into 154792ffd1c217e5c1c9774b5593a753ae553601 - [view on LGTM.com](https://lgtm.com/projects/g/Subsurface-divelog/subsurface/rev/pr-43048888356917cb4082f7d9eba49bef158531eb)

**new alerts:**

* 2 for Assignment where comparison was intended

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by djunkins** [comment]
@dirkhh It looks like LGTM doesn't like the shorthand for the bounding checks. Do you want me to change it?

**Comment by bstoeger** [comment]
I would prefer explicit if + assign. It expresses intent more clearly and is more idiomatic C.

**Comment by bstoeger** [comment]
BTW: It is surprising that two times sqrt() + atan2() + more cos() and sin() would be significantly faster than acos(). Is this gcc specific? In a quick look at the produced assembly I didn't spot any unexpected optimization.

**Comment by djunkins** [comment]
I updated the bounding checks to use an explicit 'if'.

The benchmarking results surprised me, too. I expected the simpler acos() method to be faster.

**Comment by bstoeger** [comment]
Thanks - committing so that we can move on.

**Pull request merged**